[benchmarks] overhaul benchmarks

sayakpaul opened the pull request:
# What does this PR do?

This PR considerably simplifies how we do benchmarks. Instead of using entire pipeline-level benchmarks across different tasks, we will now ONLY benchmark the diffusion network that is the most compute-intensive part in a standard diffusion workflow.

To make the estimates more realistic, we will make use of pre-trained checkpoints and dummy inputs with reasonable dimensionalities. 

I ran `benchmarking_flux.py` on an 80GB A100 on a batch size of 1 and got the following results:

<img width="1087" alt="image" src="https://github.com/user-attachments/assets/ace91098-ca81-47c4-8932-707c1e12b67e" />

Analyze the results in this Space: https://huggingface.co/spaces/diffusers/benchmark-analyzer

By default, all benchmarks will use a batch size of 1, eliminating CFG.

## How to add your benchmark?

Adding benchmarks for a new model class (`SanaTransformer2DModel`, for example) boils down to the following:

1. Define the dummy inputs of the model.
2. Define the benchmarking scenarios we should run the benchmark on.

This is what `benchmarking_flux.py` does. More modularization can be shipped afterward.

Idea would be to merge this PR with pre-configured benchmarks for a few popular models and open others to the community.

## TODOs

Utilities:

- [x] To fire the execution of the individual model-level benchmarks sequentially.
- [x] To combine CSVs from multiple different model classes.
- [x] Central dataset update and Slack notification.

@DN6 could you give the approach a quick look? I can then work on resolving the TODOs.

sayakpaul commented on benchmarks/benchmarking_utils.py:
```diff
@@ -0,0 +1,157 @@
+logger = logging.get_logger(__name__)
+
+
+def benchmark_fn(f, *args, **kwargs):
```
This automatically warms up the model. No need to do it explicitly.

sayakpaul commented on benchmarks/benchmarking_flux.py:
```diff
@@ -0,0 +1,97 @@
+
+
+if __name__ == "__main__":
+    scenarios = [
```
Covered the following scenarios:

* Regular BF16 with compilation
* NF4
* Layerwise upcasting
* Group offloading

sayakpaul commented:
Added SDXL, Wan (14B), and LTX (13B):

<details>
<summary>Results</summary>

|    |   index | scenario                                                     | model_cls                  |   num_params_M |     flops_M |   time_plain_s |   mem_plain_GB |   time_compile_s |   mem_compile_GB |   fullgraph | mode    |
|---:|--------:|:-------------------------------------------------------------|:---------------------------|---------------:|------------:|---------------:|---------------:|-----------------:|-----------------:|------------:|:--------|
|  0 |       0 | Wan-AI/Wan2.1-T2V-14B-Diffusers-bf16                         | WanTransformer3DModel      |       14288.5  | 7.85612e+08 |         10.666 |          31.17 |            9.125 |            31.77 |           1 | default |
|  1 |       1 | Wan-AI/Wan2.1-T2V-14B-Diffusers-layerwise-upcasting          | WanTransformer3DModel      |       14288.5  | 7.85612e+08 |         10.67  |          26.78 |          nan     |           nan    |         nan | nan     |
|  2 |       2 | Wan-AI/Wan2.1-T2V-14B-Diffusers-group-offload-leaf           | WanTransformer3DModel      |       14288.5  | 7.85612e+08 |         10.786 |           4.48 |          nan     |           nan    |         nan | nan     |
|  3 |       0 | stabilityai/stable-diffusion-xl-base-1.0-bf16                | UNet2DConditionModel       |        2567.46 | 5.9791e+06  |          0.088 |           5.05 |            0.058 |             5.39 |           1 | default |
|  4 |       1 | stabilityai/stable-diffusion-xl-base-1.0-layerwise-upcasting | UNet2DConditionModel       |        2567.46 | 5.9791e+06  |          0.17  |           4.89 |          nan     |           nan    |         nan | nan     |
|  5 |       2 | stabilityai/stable-diffusion-xl-base-1.0-group-offload-leaf  | UNet2DConditionModel       |        2567.46 | 5.9791e+06  |          0.386 |           0.2  |          nan     |           nan    |         nan | nan     |
|  6 |       0 | Lightricks/LTX-Video-0.9.7-dev-bf16                          | LTXVideoTransformer3DModel |       13042.6  | 1.67583e+08 |          1.439 |          25.21 |            1.134 |            25.63 |           1 | default |
|  7 |       1 | Lightricks/LTX-Video-0.9.7-dev-layerwise-upcasting           | LTXVideoTransformer3DModel |       13042.6  | 1.67583e+08 |          1.522 |          24.38 |          nan     |           nan    |         nan | nan     |
|  8 |       2 | Lightricks/LTX-Video-0.9.7-dev-group-offload-leaf            | LTXVideoTransformer3DModel |       13042.6  | 1.67583e+08 |          1.806 |           1.04 |          nan     |           nan    |         nan | nan     |

</details>

a-r-r-o-w commented:
> Cc: @a-r-r-o-w if you want to add some caching benchmarks (in a later PR), I think that would be really great!

Sounds good, I'll take it up in near future once this PR is in

sayakpaul commented on .github/workflows/benchmark.yml:
```diff
@@ -13,15 +13,15 @@ env:
       max-parallel: 1
     runs-on:
-      group: aws-g6-4xlarge-plus
+      group: aws-g6e-xlarge-plus
```
Beefier machine.

sayakpaul commented:
@DN6 a gentle ping.

sayakpaul commented:
@McPatate I have updated this PR to add a DB population script and also adjusted the workflow that will trigger it. 

I decided to create a new table because the benchmarking workflow differs from the one used in `transformers` and as such, it has served us well. So, we would like to stick to this format.

LMK what you think.

sayakpaul commented:
Okay will roll with it.

DN6 commented on .github/workflows/benchmark.yml:
```diff
@@ -3,25 +3,26 @@ name: Benchmarking tests
       max-parallel: 1
     runs-on:
-      group: aws-g6-4xlarge-plus
+      group: aws-g6e-xlarge-plus
```
Can be done in a follow up PR but it would be good to run a benchmark for more resource constrained environments too with a small GPU.

sayakpaul replied:
Indeed. Some tests might need a bigger GPU and I think we can catch and report them appropriately. Do you want to take a stab at this in a follow-up?

sayakpaul commented:
@DN6 a gentle ping.

sayakpaul commented on benchmarks/populate_into_db.py:
```diff
@@ -0,0 +1,160 @@
+
+
+def _init_benchmark(conn, branch, commit_id, commit_msg):
+    metadata = {"metadata": f"diffusers-{commit_id}"}
```
@McPatate getting: https://github.com/huggingface/diffusers/actions/runs/16043615394/job/45269816792#step:10:22

[From](https://github.com/huggingface/transformers/blob/8178c43112295bf8c4ef04c667efbbbfd34b8bca/benchmark/llama.py#L48) what I understand, I _think_ I am doing things as expected here. Could you comment?

McPatate replied:
You need to do the following:
```py
from psycopg2.extensions import register_adapter
from psycopg2.extras import Json

register_adapter(dict, Json)
```

McPatate commented on benchmarks/populate_into_db.py:
```diff
@@ -0,0 +1,160 @@
+
+
+def _init_benchmark(conn, branch, commit_id, commit_msg):
+    metadata = {"metadata": f"diffusers-{commit_id}"}
```
What is the purpose of the field? Doesn't bring much value imo, you can remove it. This field is more to store extra info, in transformers we store gpu name and model id for instance.

sayakpaul replied:
Indeed. Can we keep it null?

McPatate replied:
The field isn't nullable 😅 
Guess you can put an empty dict though.

sayakpaul commented:
Since everything is passing now, will merge this PR :) 
https://github.com/huggingface/diffusers/actions/runs/16065987231/job/45340516693

sayakpaul commented:
Also cc @a-r-r-o-w for https://github.com/huggingface/diffusers/pull/11565#issuecomment-2893942224 (not urgent, when you get time).
